File: SRP/typescript/srp_break.ts
class Product {
    public name: string;
    public price: number;

    constructor(name: string, price: number) {
        this.name = name;
        this.price = price;
    }
}

class ShoppingCart {
    private products: Product[];

    constructor() {
        this.products = [];
    }

    addProduct(product: Product): void {
        this.products.push(product);
    }

    getProducts(): Product[] {
        return this.products;
    }

    calculateTotalPrice(): number {
        let totalPrice = 0;

        for (const product of this.products) {
            totalPrice = totalPrice + product.price;
        }

        return totalPrice;
    }

    printInvoice(): void {
        console.log("\nInvoice:");

        for (const product of this.products) {
            console.log("Product Name:", product.name);
            console.log("Product Price:", product.price);
        }

        console.log("Total Price:", this.calculateTotalPrice());
    }

    saveToDb(): void {
        console.log("\nSaving shopping cart to db");
    }
}

const cart = new ShoppingCart();

const p1 = new Product("Chips", 10);
const p2 = new Product("Milk", 30);

cart.addProduct(p1);
cart.addProduct(p2);

cart.printInvoice();
cart.saveToDb();

File: SRP/java/SRPBreak.java
import java.util.ArrayList;

// Product class
class Product {
    String name;
    int price;

    public Product(String name, int price) {
        this.name = name;
        this.price = price;
    }
}

// ShoppingCart class violating SRP
class ShoppingCart {
    private ArrayList<Product> products = new ArrayList<>();

    public void addProduct(Product product) {
        products.add(product);
    }

    public ArrayList<Product> getProducts() {
        return products;
    }

    public int calculateTotalPrice() {
        int total = 0;
        for (Product product : products) {
            total += product.price;
        }
        return total;
    }

    // Violates SRP
    public void printInvoice() {
        System.out.println("\nInvoice:");
        for (Product product : products) {
            System.out.println("Product Name: " + product.name);
            System.out.println("Product Price: " + product.price);
        }
        System.out.println("Total Price: " + calculateTotalPrice());
    }

    // Violates SRP
    public void saveToDB() {
        System.out.println("\nSaving shopping cart to DB");
    }
}

public class SRPBreak {
    public static void main(String[] args) {
        ShoppingCart cart = new ShoppingCart();

        Product product1 = new Product("Chips", 10);
        Product product2 = new Product("Milk", 30);

        cart.addProduct(product1);
        cart.addProduct(product2);

        cart.printInvoice();
        cart.saveToDB();
    }
}
